File: Packages/SecuPersoData/Sources/SecuPersoData/MockDataCoordinator.swift
import Foundation
import SecuPersoDomain

public actor MockDataCoordinator {
    private let fixtureLoader: FixtureDataLoader
    private let database: EncryptedSQLiteDatabase
    private var scenario: FixtureScenario

    public init(
        fixtureLoader: FixtureDataLoader,
        database: EncryptedSQLiteDatabase,
        initialScenario: FixtureScenario = .moderate
    ) {
        self.fixtureLoader = fixtureLoader
        self.database = database
        self.scenario = initialScenario
    }

    public func setScenario(_ scenario: FixtureScenario) {
        self.scenario = scenario
    }

    public func currentScenario() -> FixtureScenario {
        scenario
    }

    public func refreshLoginEvents() throws -> [LoginEvent] {
        let events = try fixtureLoader.loadLoginEvents(for: scenario)
        try database.replaceLoginEvents(events)
        try database.appendAuditEvent("Refreshed login fixtures for scenario: \(scenario.rawValue)")
        return try database.fetchLoginEvents()
    }

    public func markLoginAsExpected(_ loginEventID: UUID) throws -> LoginEvent? {
        guard var loginEvent = try database.fetchLoginEvent(id: loginEventID) else {
            return nil
        }
        loginEvent.expected = true
        loginEvent.suspicious = false
        loginEvent.reason = "Confirmed by user"
        try database.upsertLoginEvent(loginEvent)
        try database.appendAuditEvent("User marked login event \(loginEventID.uuidString) as expected")
        return loginEvent
    }

    public func createIncident(from loginEventID: UUID) throws -> IncidentCase {
        guard let login = try database.fetchLoginEvent(id: loginEventID) else {
            throw SecuPersoDataError.loginEventNotFound(loginEventID)
        }

        let incident = IncidentCase(
            id: UUID(),
            title: "Suspicious \(login.provider.displayName) sign-in",
            severity: login.suspicious ? .high : .medium,
            createdAt: Date(),
            status: .open,
            linkedLoginEventID: login.id,
            notes: "Auto-generated from login activity.",
            resolvedAt: nil
        )

        try database.upsertIncident(incident)
        try database.appendAuditEvent("Created incident \(incident.id.uuidString)")
        return incident
    }

    public func resolveIncident(_ incidentID: UUID) throws {
        guard var incident = try database.fetchIncident(id: incidentID) else {
            throw SecuPersoDataError.incidentNotFound(incidentID)
        }

        incident.status = .resolved
        incident.resolvedAt = Date()
        try database.upsertIncident(incident)
        try database.appendAuditEvent("Resolved incident \(incidentID.uuidString)")
    }

    public func listIncidents() throws -> [IncidentCase] {
        try database.fetchIncidents()
    }

    public func loadProviderCatalog() throws -> [ProviderDescriptor] {
        let providers = try fixtureLoader.loadProviders()
        let existing = try database.fetchProviderConnections()
        let existingIDs = Set(existing.map(\.id))

        for provider in providers where !existingIDs.contains(provider.id) {
            let connection = ProviderConnection(
                id: provider.id,
                state: .disconnected,
                lastUpdatedAt: Date()
            )
            try database.upsertProviderConnection(connection)
        }

        return providers
    }

    public func updateProviderState(_ provider: ProviderID, state: ConnectionState) throws {
        let connection = ProviderConnection(
            id: provider,
            state: state,
            lastUpdatedAt: Date()
        )
        try database.upsertProviderConnection(connection)
        try database.appendAuditEvent("Provider \(provider.rawValue) moved to state \(state.rawValue)")
    }

    public func listProviderConnections() throws -> [ProviderConnection] {
        try database.fetchProviderConnections()
    }
}

File: src/data/mockDataCoordinator.ts
import type {
  ExposureRecord,
  LoginEvent,
  ProviderDescriptor,
  ProviderConnection,
  FixtureScenario,
  IncidentCase,
  ProviderID,
  ConnectionState,
} from '@domain/models';
import { providerDisplayName } from '@domain/models';
import { EncryptedSQLiteDatabase } from './encryptedDatabase';
import { FixtureDataLoader } from './fixtureDataLoader';
import { createDataError } from './errors';
import { createLocalID } from './id';

export class MockDataCoordinator {
  private fixtureLoader: FixtureDataLoader;
  private database: EncryptedSQLiteDatabase;
  private encryptFn: (data: object) => Promise<string>;
  private decryptFn: <T>(payload: string) => Promise<T>;
  private emailFingerprintFn: (email: string) => Promise<string>;
  private scenario: FixtureScenario = 'moderate';

  constructor(
    fixtureLoader: FixtureDataLoader,
    database: EncryptedSQLiteDatabase,
    encryptFn: (data: object) => Promise<string>,
    decryptFn: <T>(payload: string) => Promise<T>,
    emailFingerprintFn: (email: string) => Promise<string>,
    initialScenario: FixtureScenario = 'moderate',
  ) {
    this.fixtureLoader = fixtureLoader;
    this.database = database;
    this.encryptFn = encryptFn;
    this.decryptFn = decryptFn;
    this.emailFingerprintFn = emailFingerprintFn;
    this.scenario = initialScenario;
  }

  setScenario(scenario: FixtureScenario): void {
    this.scenario = scenario;
  }

  currentScenario(): FixtureScenario {
    return this.scenario;
  }

  async refreshLoginEvents(): Promise<LoginEvent[]> {
    const events = this.fixtureLoader.loadLoginEvents(this.scenario);
    await this.database.replaceLoginEvents(events, this.encryptFn);
    await this.database.appendAuditEvent(`Refreshed login fixtures for scenario: ${this.scenario}`);
    return this.database.fetchLoginEvents(this.decryptFn);
  }

  async refreshExposures(): Promise<ExposureRecord[]> {
    const exposures = this.fixtureLoader.loadExposures(this.scenario)
    await this.database.replaceExposures(exposures, this.encryptFn)
    await this.database.appendAuditEvent(`Refreshed exposure fixtures for scenario: ${this.scenario}`)
    return this.database.fetchExposures(this.decryptFn)
  }

  async markLoginAsExpected(loginEventID: string): Promise<LoginEvent | null> {
    const loginEvent = await this.database.fetchLoginEvent(loginEventID, this.decryptFn);
    if (!loginEvent) return null;

    const updated: LoginEvent = {
      ...loginEvent,
      expected: true,
      suspicious: false,
      reason: 'Confirmed by user',
    };
    await this.database.upsertLoginEvent(updated, this.encryptFn);
    await this.database.appendAuditEvent(`User marked login event ${loginEventID} as expected`);
    return updated;
  }

  async createIncident(loginEventID: string): Promise<IncidentCase> {
    const login = await this.database.fetchLoginEvent(loginEventID, this.decryptFn);
    if (!login) throw createDataError('loginEventNotFound', loginEventID);

    const incident: IncidentCase = {
      id: createLocalID('incident'),
      title: `Suspicious ${providerDisplayName(login.provider)} sign-in`,
      severity: login.suspicious ? 'high' : 'medium',
      createdAt: new Date().toISOString(),
      status: 'open',
      linkedLoginEventID: login.id,
      notes: 'Auto-generated from login activity.',
      resolvedAt: null,
    };

    await this.database.upsertIncident(incident, this.encryptFn);
    await this.database.appendAuditEvent(`Created incident ${incident.id}`);
    return incident;
  }

  async resolveIncident(incidentID: string): Promise<void> {
    const incident = await this.database.fetchIncident(incidentID, this.decryptFn);
    if (!incident) throw createDataError('incidentNotFound', incidentID);

    const resolved: IncidentCase = {
      ...incident,
      status: 'resolved',
      resolvedAt: new Date().toISOString(),
    };
    await this.database.upsertIncident(resolved, this.encryptFn);
    await this.database.appendAuditEvent(`Resolved incident ${incidentID}`);
  }

  async listIncidents(): Promise<IncidentCase[]> {
    return this.database.fetchIncidents(this.decryptFn);
  }

  async loadProviderCatalog(): Promise<ProviderDescriptor[]> {
    return this.fixtureLoader.loadProviders();
  }

  async updateProviderState(provider: ProviderID, state: ConnectionState): Promise<void> {
    const connection: ProviderConnection = {
      id: provider,
      state,
      lastUpdatedAt: new Date().toISOString(),
    };
    await this.database.upsertProviderConnection(connection, this.encryptFn);
  }

  async listProviderConnections(): Promise<ProviderConnection[]> {
    const connections = await this.database.fetchProviderConnections(this.decryptFn);
    const existingIDs = new Set(connections.map((c) => c.id));
    for (const id of ['google', 'outlook', 'other'] as ProviderID[]) {
      if (!existingIDs.has(id)) {
        const defaultConnection: ProviderConnection = {
          id,
          state: 'disconnected',
          lastUpdatedAt: new Date().toISOString(),
        };
        await this.database.upsertProviderConnection(defaultConnection, this.encryptFn);
      }
    }
    return this.database.fetchProviderConnections(this.decryptFn);
  }
}
